### PR TITLE
more performance in explosion :)

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by robske110** [comment]
Could you pleae document what performance improvements were made? I can only see type hints

File: src/pocketmine/level/Explosion.php [hunk 1]
```diff
@@ -1,85 +1,77 @@
...
+use pocketmine\network\Network;
 use pocketmine\network\protocol\ExplodePacket;
-
+use pocketmine\Server;
 use pocketmine\utils\Random;
-
-class Explosion{
-
-	private $rays = 16; //Rays
+class Explosion extends Level{ //implements vector iterator
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:23]: Vector iterator -> Was added by me once

**Comment by inxomnyaa** [comment]
No performance improvements. Would say, as of testing, even lags more.

File: src/pocketmine/level/Explosion.php (hunk 1 above)
Comment (inxomnyaa) [reply src/pocketmine/level/Explosion.php:23]: Why would you even need Level?

File: src/pocketmine/level/Explosion.php [hunk 2]
```diff
@@ -1,85 +1,77 @@
...
 use pocketmine\entity\Entity;
 use pocketmine\event\entity\EntityDamageByBlockEvent;
 use pocketmine\event\entity\EntityDamageByEntityEvent;
 use pocketmine\event\entity\EntityDamageEvent;
 use pocketmine\event\entity\EntityExplodeEvent;
-use pocketmine\event\block\BlockUpdateEvent;
 use pocketmine\item\Item;
 use pocketmine\math\AxisAlignedBB;
 use pocketmine\math\Math;
+use pocketmine\event\block\BlockUpdateEvent;
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:12]: Why the hell would you move this out of the alphabetic order?!

File: src/pocketmine/level/Explosion.php [hunk 3]
```diff
@@ -1,85 +1,77 @@
...
 use pocketmine\math\Vector3;
 use pocketmine\nbt\tag\ByteTag;
 use pocketmine\nbt\tag\CompoundTag;
 use pocketmine\nbt\tag\DoubleTag;
 use pocketmine\nbt\tag\ListTag;
 use pocketmine\nbt\tag\FloatTag;
+use pocketmine\network\Network;
 use pocketmine\network\protocol\ExplodePacket;
-
+use pocketmine\Server;
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:21]: unneeded. all you used was getServer, no need to import

File: src/pocketmine/level/Explosion.php [hunk 4]
```diff
@@ -1,85 +1,77 @@
...
 		$this->size = max($size, 0);
 		$this->what = $what;
 	}
-
-	/**
-	 * @return bool
-	 */
+	public function explode(){
+		if($this->explodeA()){
+			return $this->explodeB();
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:39]: why? Doesn't makes sence in my eyes. if explode a use explode b too?

File: src/pocketmine/level/Explosion.php [hunk 5]
```diff
@@ -1,85 +1,77 @@
...
-						$pointerZ = $this->source->z;
-
-						for($blastForce = $this->size * (mt_rand(700, 1300) / 1000); $blastForce > 0; $blastForce -= $this->stepLen * 0.75){
-							$x = (int) $pointerX;
-							$y = (int) $pointerY;
-							$z = (int) $pointerZ;
+		for((int) $i = 0; $i < $this->rays; ++$i){
+			for((int) $j = 0; $j < $this->rays; ++$j){
+				for((int) $k = 0; $k < $this->rays; ++$k){
+					if($i == 0 || $i == $mRays || $j == 0 || $j == $mRays || $k == 0 || $k == $mRays){
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:53]: Still, no performance improvements. PHP Preprocessor ftw.

File: src/pocketmine/level/Explosion.php [hunk 6]
```diff
@@ -1,85 +1,77 @@
...
+						$vector->setComponents((double) $i / (double) $mRays * 2 - 1, (double) $j / (double) $mRays * 2 - 1, (double) $k / (double) $mRays * 2 - 1);
+						(double) $len = $vector->length();
+						$vector->setComponents(($vector->x / $len) * $this->stepLen, ($vector->y / $len) * $this->stepLen, ($vector->z / $len) * $this->stepLen);
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:56]: may be the only change

File: src/pocketmine/level/Explosion.php [hunk 7]
```diff
@@ -92,17 +84,20 @@ public function explodeA(){
...
-
+		(double) $yield = (1 / $this->size) * 100;
+		(double) $explosionSize = $this->size * 2;
+		(double) $minX = Math::floorFloat($this->source->x - $explosionSize - 1);
+		(double) $maxX = Math::ceilFloat($this->source->x + $explosionSize + 1);
+		(double) $minY = Math::floorFloat($this->source->y - $explosionSize - 1);
+		(double) $maxY = Math::ceilFloat($this->source->y + $explosionSize + 1);
+		(double) $minZ = Math::floorFloat($this->source->z - $explosionSize - 1);
+		(double) $maxZ = Math::ceilFloat($this->source->z + $explosionSize + 1);
+		$explosionBB = new AxisAlignedBB($minX, $minY, $minZ, $maxX, $maxY, $maxZ);
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:100]: Well, thats even more code

File: src/pocketmine/level/Explosion.php [hunk 8]
```diff
@@ -92,17 +84,20 @@ public function explodeA(){
...
 		if($this->what instanceof Entity){
 			$this->level->getServer()->getPluginManager()->callEvent($ev = new EntityExplodeEvent($this->what, $this->source, $this->affectedBlocks, $yield));
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:102]: Todo: for gamerules and entity ai

File: src/pocketmine/level/Explosion.php [hunk 9]
```diff
@@ -112,48 +107,30 @@ public function explodeB(){
...
-		$explosionSize = $this->size * 2;
-		$minX = Math::floorFloat($this->source->x - $explosionSize - 1);
-		$maxX = Math::ceilFloat($this->source->x + $explosionSize + 1);
-		$minY = Math::floorFloat($this->source->y - $explosionSize - 1);
-		$maxY = Math::ceilFloat($this->source->y + $explosionSize + 1);
-		$minZ = Math::floorFloat($this->source->z - $explosionSize - 1);
-		$maxZ = Math::ceilFloat($this->source->z + $explosionSize + 1);
-
-		$explosionBB = new AxisAlignedBB($minX, $minY, $minZ, $maxX, $maxY, $maxZ);
-
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:125]: I see. moved it up.

File: src/pocketmine/level/Explosion.php [hunk 10]
```diff
@@ -112,48 +107,30 @@ public function explodeB(){
...
+			(double) $distance = $entity->distance($this->source) / $explosionSize;
 			if($distance <= 1){
 				$motion = $entity->subtract($this->source)->normalize();
-
-				$impact = (1 - $distance) * ($exposure = 1);
-
-				$damage = (int) ((($impact * $impact + $impact) / 2) * 8 * $explosionSize + 1);
-
+				(double) $impact = (1 - $distance) * ($exposure = 1);
+				(int) $exposure = 1;
```
Comment (inxomnyaa) [inline src/pocketmine/level/Explosion.php:116]: Why resetting this?

**Comment by DaPigGuy** [comment]
:-1:

**Comment by Aericio** [comment]
NycuRO's alt account, you can hate me later Nycu <3

**Comment by dktapps** [comment]
@Aericio Thanks for releasing that information, the ban button will be hit at Genisys

**Comment by inxomnyaa** [comment]
No difference, 10000 block tnt testing prooved it.

**Comment by daniktheboss** [comment]
Yup

**Pull request closed without merging**